website/blog/2022-05-24-joining-snapshot-complexity.md

```diff
@@ -0,0 +1,288 @@
+
+```sql
+build_spine as (
+	historical_table_1.*,
+	historical_table_2.product_order_id,
+	historical_table_3.other_entity_grain_id, 
+	...
+from historical_table_1
+left join
+	historical_table_2 
+	on historical_table_1.product_id = historical_table_2.product_id
+left join 
+	historical_table_3
+	on historical_table_1.product_id = historical_table_3.product_id
+... )
+```
+
+The result will be all the columns from your first table, fanned out as much as possible by the added id columns. We will use these id columns to join the historical data from our tables. 
+
+| product_id | important_status | dbt_valid_from | dbt_valid_to | product_order_id |
+| --- | --- | --- | --- | --- |
+| 1 | available | 2021-11-15 16:00:0000 | NULL | 1A |
+| 1 | pending | 2021-11-10 08:00:000 | 2021-11-15 16:00:0000 | 1A |
+| 1 | not_available | 2021-10-01 10:00:000 | 2021-11-10 08:00:000 | 1A |
+| 1 | available | 2021-11-15 16:00:0000 | NULL | 1B |
+| 1 | pending | 2021-11-10 08:00:000 | 2021-11-15 16:00:0000 | 1B |
+| 1 | not_available | 2021-10-01 10:00:000 | 2021-11-10 08:00:000 | 1B |
+| 2 | not_available | 2021-11-15 15:30:0000 | NULL  | 2C |
+
+## Step 5: Join your snapshots onto id spine
+
+:::important What happens in this step? 
+Step 5 walks you through the logic of the snapshot join, and the macro that will make the joins simpler.
+:::
+
+Now, I’m going to recommend you build individual CTEs with one join at a time. Why do we build a CTE with a single join, rather than all the joins in one? So many reasons, but there are two big ones.
+
+- **First**, this is complicated. You will need to troubleshoot, and the easiest way to enable troubleshooting is to separate your join logic in individual CTEs. By building your code this way, you can easily throw a `select * from last_cte` to check that your logic is doing what you think it should be doing before adding more complex joins.
+- **Second**, you are using the `valid_from` and `valid_to` values of each newly joined table to determine the new `valid_from` and `valid_to` timestamps for the resulting table – where both rows are valid within the same timespans. While you could accomplish this in one big massive join, it will become very complex and difficult to troubleshoot when you run into funky results.
+
+### **<Term id="dry">DRY</Term> — it’s macro time!**
+
+This macro finishes your join CTE, which allows you to add columns from the new table you’re joining before calling the macro. It also assumes you’ve replaced your `valid_to = NULL` with an actual date type with an actual date that indicates a row is currently valid.
+
+Your parameters are `cte_join`, the table that is creating the spine of your final model, `cte_join_on`, which is the new table you’re joining onto the spine. The *`valid_to`* and `valid_from` values for both of these tables, and the ids onto which you are joining (named twice in case they have different column names, but in most instances these two parameters will be the same name).
```

Comment:
Is the `valid_to` in the second sentence meant to be italicized?

Reply:
Not sure, but I'll remove it